### PR TITLE
add support for index-signature to serializable object

### PR DESCRIPTION
**Pull request by Kelerchian**
As I was working on SW's case, I notice that some data TypeScript's Record is not supported by SerializableObject because `{ [string]: unknown }` is treated differently from `Record<string, unknown>`. This PR adds support to it.

Currently, the code below doesn't work
```
designEvent("x").withPayload<{ someField: Record<string, string> }>()
```

### CODE REVIEW

File: machine-runner/src/utils/type-utils.ts [hunk 1]
```diff
@@ -23,12 +23,12 @@ export type NotAnyOrUnknown<T> = any extends T ? never : T
 
 type SerializablePrimitive = number | string | boolean | null | undefined
 
+type SerializableRecord = Record<string, SerializablePrimitive>
```
Comment (rkuhn) [inline machine-runner/src/utils/type-utils.ts:26]: Does this mean that `{ [_: string]: Record<string, number[]> }` is not considered serialisable? Or is that covered via SerializableObject?

**Review by rkuhn — APPROVED** [review]
Looking good, just one question.

File: machine-runner/src/utils/type-utils.ts (hunk 1 above)
Comment (Kelerchian) [reply machine-runner/src/utils/type-utils.ts:26]: You're right. It should have been SerializableObject here.

Comment (Kelerchian) [reply machine-runner/src/utils/type-utils.ts]: Fixed in https://github.com/Actyx/machines/commit/4e14c47119e5bfaffb3479731fb8241b2d7dffef

File: machine-runner/src/utils/type-utils.ts [hunk 2]
```diff
@@ -21,14 +21,18 @@ export type Expect<T extends true> = T
...
-type SerializablePrimitive = number | string | boolean | null | undefined
+type SerializableValue =
+  | string
+  | number
+  | boolean
+  | null
+  | SerializableObject
+  | SerializableArray
+  // Record type cannot be circular https://github.com/microsoft/TypeScript/issues/41164
+  | Record<string, unknown>
```
[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/utils/type-utils.ts:32]: or could https://github.com/microsoft/TypeScript/issues/41164#issuecomment-1347389099 help?

[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/utils/type-utils.ts:32]:
```
withPayload<{
  c: Record<string, string>
}>
```

This will fail

**Review by rkuhn — COMMENTED** [review]
Basically looking good, but it would be better if the code contained better comments explaining the constraints of type-checking (i.e. what doesn’t work).

File: machine-runner/src/utils/type-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/utils/type-utils.ts:32]: > or could [microsoft/TypeScript#41164 (comment)](https://github.com/microsoft/TypeScript/issues/41164#issuecomment-1347389099) help?

But there is no way to extend number, string, and other primitives. We need ThisWorks that also contain primitives

**Comment by Kelerchian** [comment]
> Basically looking good, but it would be better if the code contained better comments explaining the constraints of type-checking (i.e. what doesn’t work).

Comments added in https://github.com/Actyx/machines/pull/44/commits/9686c0e1f37e8b7057033f3b2499cf70a7283290

File: machine-runner/src/design/event.ts [hunk 3]
```diff
@@ -79,10 +80,11 @@ export namespace MachineEvent {
...
      * Declares the payload type for this MachineEvent.
      *
-     * Event payload will be serialized. See that the payload definition put into
-     * `withPayload` method does not include non-serializable properties such as
-     * symbol, Function, Date, BigInt. machine-runner cannot put this constraint
-     * into the type definition due to constraint of the TypeScript compiler.
+     * Event payload will be serialized. Payload definition cannot have fields
+     * that cannot be serialized and deserialized back via JSON.stringify and
+     * JSON.parse. See that the payload definition type argument does not include
+     * non-serializable properties, especially `symbol` and `number` indexed
```
Comment (rkuhn) [inline machine-runner/src/design/event.ts:86]: are you sure that `number` indexing is problematic?

**Review by rkuhn — APPROVED** [review]
Looking good, just one question for two words in the docs.

File: machine-runner/src/design/event.ts (hunk 3 above)
Comment (Kelerchian) [reply machine-runner/src/design/event.ts:86]: I am not sure. Let me check.

Comment (Kelerchian) [reply machine-runner/src/design/event.ts:86]: No it is not problematic. Revising my answer

**Review by rkuhn — APPROVED** [review]
Great work, now it looks water-tight!

**Pull request merged**